src/extension.ts

```diff
@@ -107,146 +142,57 @@ async function activeServer(context: ExtensionContext, document: TextDocument) {
   if (clients.has(clientsKey)) {
     return;
   }
-
-  const currentWorkingDir = folder ? folder.uri.fsPath : path.dirname(uri.fsPath);
-
   // Set the key to null to prevent multiple servers being launched at once
   clients.set(clientsKey, null);
 
-  const logLevel = workspace.getConfiguration('haskell', uri).trace.server;
-  const clientLogLevel = workspace.getConfiguration('haskell', uri).trace.client;
-  const logFile: string = workspace.getConfiguration('haskell', uri).logFile;
+  const config = initConfig(workspace.getConfiguration('haskell', uri), uri, folder);
+  const logger: Logger = initLoggerFromConfig(config);
 
-  const outputChannel: OutputChannel = window.createOutputChannel(langName);
+  logConfig(logger, config);
 
-  const logFilePath = logFile !== '' ? path.resolve(currentWorkingDir, expandHomeDir(logFile)) : undefined;
-  const logger: Logger = new ExtensionLogger('client', clientLogLevel, outputChannel, logFilePath);
-  if (logFilePath) {
-    logger.info(`Writing client log to file ${logFilePath}`);
-  }
-  logger.log('Environment variables:');
-  Object.entries(process.env).forEach(([key, value]: [string, string | undefined]) => {
-    // only list environment variables that we actually care about.
-    // this makes it safe for users to just paste the logs to whoever,
-    // and avoids leaking secrets.
-    if (['PATH'].includes(key)) {
-      logger.log(`  ${key}: ${value}`);
-    }
-  });
-
-  let serverExecutable: string;
-  let addInternalServerPath: string | undefined; // if we download HLS, add that bin dir to PATH
+  let hlsExecutable: HlsExecutable;
   try {
-    [serverExecutable, addInternalServerPath] = await findHaskellLanguageServer(
-      context,
-      logger,
-      currentWorkingDir,
-      folder,
-    );
-    if (!serverExecutable) {
-      return;
-    }
+    hlsExecutable = await findHaskellLanguageServer(context, logger, config.ghcupConfig, config.workingDir, folder);
   } catch (e) {
-    if (e instanceof MissingToolError) {
-      const link = e.installLink();
-      if (link) {
-        if (await window.showErrorMessage(e.message, `Install ${e.tool}`)) {
-          env.openExternal(link);
-        }
-      } else {
-        await window.showErrorMessage(e.message);
-      }
-    } else if (e instanceof HlsError) {
-      logger.error(`General HlsError: ${e.message}`);
-      window.showErrorMessage(e.message);
-    } else if (e instanceof NoMatchingHls) {
-      const link = e.docLink();
-      logger.error(`${e.message}`);
-      if (await window.showErrorMessage(e.message, 'Open documentation')) {
-        env.openExternal(link);
-      }
-    } else if (e instanceof Error) {
-      logger.error(`Internal Error: ${e.message}`);
-      window.showErrorMessage(e.message);
-    }
-    if (e instanceof Error) {
-      // general stack trace printing
-      if (e.stack) {
-        logger.error(`${e.stack}`);
-      }
-    }
+    await handleInitializationError(e, logger);
```

Comment:
```suggestion
    await handleInitializationError(e, logger);
    clients.delete(clientsKey);
```
otherwise slot will be forever taken